src/Typedefs/TermParse.idr

```diff
@@ -107,28 +113,128 @@ injection (FS FZ) [a, b]             x = Right x
+injProd [(_ ** x), (_ ** y)] = (x, y)
+injProd ((_ ** x) :: y :: z :: zs) = (x, assert_total $ injProd (y :: z :: zs))
+
+-- we need to check all keys are in increasing order and of the format _X from 0 to l
+parseVect : List (String, JSON) -> (l : Nat) -> JSONM (Vect l JSON)
+parseVect ls n = case decEq (length ls) n of
+                      Yes prf => rewrite sym prf in pure $ map snd (fromList ls)
+                      No _ => Left ("incompatible lengths: " ++ show (length ls)
+                                    ++ " and " ++ show n)
+
+deserialiseJSON : (t : TDefR n) -> (ts : Vect n (a ** (JSON -> JSONM a)))
+               -> JSON -> Either String (Ty (map DPair.fst ts) t)
+deserialiseJSON T0 ts json = Left "parsing 0"
+deserialiseJSON T1 ts json = pure ()
+deserialiseJSON td@(TSum {k = k} tds) ts json =  do
+  [(key, obj)] <- expectObject json
+    | _ => Left "Object doens't contain exactly 1 element"
+  key' <- parseKey key
+  t' <- deserialiseJSON (assert_smaller td (index key' tds)) ts obj
+  pure $ injection key' tds t'
+deserialiseJSON (TProd products {k}) ts json {n} = do
+  keyValues <- expectObject json
+  vec <- parseVect keyValues (2 + k)
+  res <- the (JSONM $ Vect (2 + k) (td : TDefR n ** Ty (map DPair.fst ts) td))
+             (traverse (\(def, j) => MkDPair def <$> (assert_total $ deserialiseJSON def ts j))
+                       (zip products vec))
+  -- Alright explaination time:
+  --
+  -- In the previous line we compute a vector of parsed JSON along with the TDef they are supposed
+  -- to be a member of. This dependent pair comes from zipping `products` and the vector of JSONs
+  -- and then traversing it while parsing the jsons using the TDefs as a schema to say "here is
+  -- the expected type this value shoud parse as".
+  -- However in order to return the Idris type associated with this product type we need a proof
+  -- that the values we parsed (and their type) correspond to the types we expect from `products`
+  -- (which itself is the vector of TDefs that  defines which types we expect in which order).
+  --
+  -- But now we're stuck because we have a vector of (def : TDefR n ** f def) where the first
+  -- projection _is_ the TDef from `products` (because we zipped them and took the first element to
+  -- make the returning DPair), but we don't have a proof of it.
+  --
+  -- I tried using `let prf = the (map fst res = products) (believe_me Refl)` and rewriting but
+  -- it just wouldn't work. In the end, since as a programmer I _know_ this is true because of the
+  -- implementation, I decided to force the hand of the compiler and use `belive_me` at the call
+  -- site of `injprod` (which builds the Idris value from the given product) rather than in a
+  -- rewrite. This is safe but dangerous if we ever change the line right above this scary comment.
+  --
+  -- Please don't ignore this comment when you change the code around those two lines.
+  pure $ believe_me $ injProd res
```

Comment:
It's hard to say without digging deeper, but intuitively we probably need a custom function instead of `traverse` + `zip`. Let's make a separate issue about this maybe?

Reply:
I'll make an issue about this